File: C#/0438. Find All Anagrams in a String.cs
public class Solution {
    public IList<int> FindAnagrams(string s, string p) {
        IList<int> rep=new List<int>();
        if(s.Length<p.Length){
            return rep;
        }
        Dictionary<char,int> sDict=new Dictionary<char,int>();
        Dictionary<char,int> pDict=new Dictionary<char,int>();
        for(int i=0;i<p.Length;i++){
            pDict[p[i]]=!pDict.ContainsKey(p[i])?1:pDict[p[i]]+1;
            sDict[s[i]]=!sDict.ContainsKey(s[i])?1:sDict[s[i]]+1;
        }
        if(IsAnagram(sDict,pDict)){
                rep.Add(0);
        }
        for(int i=1;i<s.Length-p.Length+1;i++){
            sDict[s[i-1]]-=1;
            if(sDict[s[i-1]]==0){
                sDict.Remove(s[i-1]);
            }
            sDict[s[i+p.Length-1]]=!sDict.ContainsKey(s[i+p.Length-1])?1:(sDict[s[i+p.Length-1]]+1);
            if(IsAnagram(sDict,pDict)){
                rep.Add(i);
            }  
        }
        return rep; 
    }
    public bool IsAnagram(Dictionary<char,int> sDict,Dictionary<char,int> pDict){
        bool flag=true;
        foreach(KeyValuePair<char,int> kvp in sDict){
            if(!pDict.ContainsKey(kvp.Key) || pDict[kvp.Key]!=kvp.Value){
                flag=false;
                break;
            }
        }
        return flag;
    }
}

File: archives/20190519python/0438. Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        rep=[]
        if len(p)>len(s):
            return rep
        
        dict_s={}
        for c in s[:len(p)]:
            if c not in dict_s:
                dict_s[c]=1
            else:
                dict_s[c]+=1
        
        dict_p={}
        for c in p:
            if c not in dict_p:
                dict_p[c]=1
            else:
                dict_p[c]+=1
        
        if self.isAnagrams(dict_s,dict_p):
            rep.append(0)
        
        for i in range(1,len(s)-len(p)+1):
            dict_s[s[i-1]]-=1
            if s[i+len(p)-1] not in dict_s:
                dict_s[s[i+len(p)-1]]=1
            else:
                dict_s[s[i+len(p)-1]]+=1
                
            if self.isAnagrams(dict_s,dict_p):
                rep.append(i)
        return rep
    
    
    
    def isAnagrams(self,dicts,dictp):
        for k in dictp:
            if k not in dicts or dicts[k]!=dictp[k]:
                return False
        return True
        
                
        
